pac4j-kerberos/src/main/java/org/pac4j/kerberos/credentials/authenticator/SunJaasKerberosTicketValidator.java

```diff
@@ -0,0 +1,251 @@
+package org.pac4j.kerberos.credentials.authenticator;
+
+import org.ietf.jgss.*;
+import org.pac4j.core.exception.BadCredentialsException;
+import org.pac4j.core.exception.TechnicalException;
+import org.pac4j.core.util.CommonHelper;
+import org.pac4j.core.util.InitializableObject;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.core.io.Resource;
+
+import javax.security.auth.Subject;
+import javax.security.auth.kerberos.KerberosPrincipal;
+import javax.security.auth.login.AppConfigurationEntry;
+import javax.security.auth.login.Configuration;
+import javax.security.auth.login.LoginContext;
+import javax.security.auth.login.LoginException;
+import java.io.IOException;
+import java.security.Principal;
+import java.security.PrivilegedActionException;
+import java.security.PrivilegedExceptionAction;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * Implementation of {@link KerberosTicketValidator} which uses the SUN JAAS
+ * login module, which is included in the SUN JRE, it will not work with an IBM JRE.
+ * The whole configuration is done in this class, no additional JAAS configuration
+ * is needed.
+ *
+ * @author Garry Boyce
+ * @since 2.1.0
+ * <p>
+ * originally from spring-kerberos project
+ */
+public class SunJaasKerberosTicketValidator extends InitializableObject implements KerberosTicketValidator {
+
+    private String servicePrincipal;
+    private Resource keyTabLocation;
+    private Subject serviceSubject;
+    private boolean holdOnToGSSContext;
+    private boolean debug = false;
+    protected final Logger logger = LoggerFactory.getLogger(getClass());
+
+    @Override
+    public KerberosTicketValidation validateTicket(byte[] token) throws BadCredentialsException {
+        try {
+            return Subject.doAs(this.serviceSubject, new KerberosValidateAction(token));
+        } catch (PrivilegedActionException e) {
+            throw new BadCredentialsException("Kerberos validation not successful", e);
+        }
+    }
+
+    @Override
+    protected void internalInit() {
+        try {
```

Comment:
We should call the `init()` method first, instead of letting the `KerberosAuthenticator` do that for us.